## Data Wrangling

In [1]:
import re 
import glob
import zipfile
import pandas as pd
from bs4 import BeautifulSoup

This data set includes information about individual rides made in a bike-sharing system covering the greater San Francisco Bay area.
https://www.lyft.com/bikes/bay-wheels/system-data

The Data
Each trip is anonymized and includes:

* Trip Duration (seconds)
* Start Time and Date
* End Time and Date
* Start Station ID
* Start Station Name
* Start Station Latitude
* Start Station Longitude
* End Station ID
* End Station Name
* End Station Latitude
* End Station Longitude
* Bike ID
* User Type (Subscriber or Customer – “Subscriber” = Member or “Customer” = Casual)

In [2]:
with open('./data/content.html', 'r') as f:
    contents = f.read()
    soup = BeautifulSoup(contents, 'lxml')
    print(soup)

<html><head>
<!--

  Amazon S3 Bucket listing.


  Copyright (C) 2008 Francesco Pasqualini

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 3 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.

  -->
<!--

  Modified by Nolan Lawson!  (http://nolanlawson.com).  I'm keeping the spirit of the
  GPL alive by issuing this with the same license!

  -->
<title>Index of bucket "baywheels-data"</title>
<link href="//netdna.bootstrapc

In [3]:
# Make a list containing all the zip-file names
zip_list = glob.glob('./data/*.zip')
zip_list                          

['./data\\2017-fordgobike-tripdata.csv.zip',
 './data\\201801-fordgobike-tripdata.csv.zip',
 './data\\201802-fordgobike-tripdata.csv.zip',
 './data\\201803-fordgobike-tripdata.csv.zip',
 './data\\201804-fordgobike-tripdata.csv.zip',
 './data\\201805-fordgobike-tripdata.csv.zip',
 './data\\201806-fordgobike-tripdata.csv.zip',
 './data\\201807-fordgobike-tripdata.csv.zip',
 './data\\201808-fordgobike-tripdata.csv.zip',
 './data\\201809-fordgobike-tripdata.csv.zip',
 './data\\201810-fordgobike-tripdata.csv.zip',
 './data\\201811-fordgobike-tripdata.csv.zip',
 './data\\201812-fordgobike-tripdata.csv.zip',
 './data\\201901-fordgobike-tripdata.csv.zip',
 './data\\201902-fordgobike-tripdata.csv.zip',
 './data\\201903-fordgobike-tripdata.csv.zip',
 './data\\201904-fordgobike-tripdata.csv.zip',
 './data\\201905-baywheels-tripdata.csv.zip',
 './data\\201906-baywheels-tripdata.csv.zip',
 './data\\201907-baywheels-tripdata.csv.zip',
 './data\\201908-baywheels-tripdata.csv.zip',
 './data\\201909-ba

In [4]:
# Extract all zipfile using zipfile library
for zip_file in zip_list:
    with zipfile.ZipFile(zip_file, 'r') as zip_ref:
        zip_ref.extractall("./data/")

In [4]:
csv_list = glob.glob('./data/*.csv')
csv_list      

['./data\\2017-fordgobike-tripdata.csv',
 './data\\201801-fordgobike-tripdata.csv',
 './data\\201802-fordgobike-tripdata.csv',
 './data\\201803-fordgobike-tripdata.csv',
 './data\\201804-fordgobike-tripdata.csv',
 './data\\201805-fordgobike-tripdata.csv',
 './data\\201806-fordgobike-tripdata.csv',
 './data\\201807-fordgobike-tripdata.csv',
 './data\\201808-fordgobike-tripdata.csv',
 './data\\201809-fordgobike-tripdata.csv',
 './data\\201810-fordgobike-tripdata.csv',
 './data\\201811-fordgobike-tripdata.csv',
 './data\\201812-fordgobike-tripdata.csv',
 './data\\201901-fordgobike-tripdata.csv',
 './data\\201902-fordgobike-tripdata.csv',
 './data\\201903-fordgobike-tripdata.csv',
 './data\\201904-fordgobike-tripdata.csv',
 './data\\201905-baywheels-tripdata.csv',
 './data\\201906-baywheels-tripdata.csv',
 './data\\201907-baywheels-tripdata.csv',
 './data\\201908-baywheels-tripdata.csv',
 './data\\201909-baywheels-tripdata.csv',
 './data\\201910-baywheels-tripdata.csv',
 './data\\201911-ba

# Data Assesment

### 2017 Dataframe

In [6]:
df_2017 = pd.read_csv('./data/2017-fordgobike-tripdata.csv')
df_2017

duration_sec                start_time                  end_time  \
0              80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1              78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
2              45768  2017-12-31 22:45:48.4110  2018-01-01 11:28:36.8830   
3              62172  2017-12-31 17:31:10.6360  2018-01-01 10:47:23.5310   
4              43603  2017-12-31 14:23:14.0010  2018-01-01 02:29:57.5710   
...              ...                       ...                       ...   
519695           435  2017-06-28 10:00:54.5280  2017-06-28 10:08:10.4380   
519696           431  2017-06-28 09:56:39.6310  2017-06-28 10:03:51.0900   
519697           424  2017-06-28 09:47:36.3470  2017-06-28 09:54:41.1870   
519698           366  2017-06-28 09:47:41.6640  2017-06-28 09:53:47.7150   
519699           188  2017-06-28 09:49:46.3770  2017-06-28 09:52:55.3380   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
4                    239                      Bancroft Way at Telegraph Ave   
...                  ...                                                ...   
519695                81                                 Berry St at 4th St   
519696                66                              3rd St at Townsend St   
519697                21   Montgomery St BART Station (Market St at 2nd St)   
519698                58                               Market St at 10th St   
519699                25                                Howard St at 2nd St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
4                    37.868813              -122.258764             247   
...                        ...                      ...             ...   
519695               37.775880              -122.393170              45   
519696               37.778742              -122.392741             321   
519697               37.789625              -122.400811              48   
519698               37.776619              -122.417385              59   
519699               37.787522              -122.397405              48   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
4                               Fulton St at Bancroft Way   
...                                                   ...   
519695                                5th St at Howard St   
519696                                      5th at Folsom   
519697                                2nd St at S Park St   
519698                        S Van Ness Ave at Market St   
519699                                2nd St at S Park St   

        end_station_latitude  end_station_longitude  bike_id   user_type  
0                  37.778768            -122.415929       96    Customer  
1                  37.766210            -122.426614       88    Customer  
2                  37.870348            -122.267764     1094    Customer  
3                  37.783899            -122.408445     2831    Customer  
4                  37.867789            -122.265896     3167  Subscriber  
...                      ...                    ...      ...  

In [46]:
df_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 519700 entries, 0 to 519699
Data columns (total 13 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   duration_sec             519700 non-null  int64  
 1   start_time               519700 non-null  object 
 2   end_time                 519700 non-null  object 
 3   start_station_id         519700 non-null  int64  
 4   start_station_name       519700 non-null  object 
 5   start_station_latitude   519700 non-null  float64
 6   start_station_longitude  519700 non-null  float64
 7   end_station_id           519700 non-null  int64  
 8   end_station_name         519700 non-null  object 
 9   end_station_latitude     519700 non-null  float64
 10  end_station_longitude    519700 non-null  float64
 11  bike_id                  519700 non-null  int64  
 12  user_type                519700 non-null  object 
dtypes: float64(4), int64(4), object(5)
memory usage: 51.5+ MB


### 2018 Dataframes

In [76]:
# Function to check column equality for csv in one year 
def column_check(index):
#     diff = []
    columns = pd.read_csv(csv_list[index]).columns
    for i in csv_list[index:index+12]:
        df = pd.read_csv(i)
        if columns.all() == df.columns.all():
            print(i, 'columns is identical')
        else:
            print(i, 'columns is NOT identical')
#             j = i
#         return diff.append() 

In [67]:
# Column check for year 2018, starterd from index 1 to 12
column_check(1)

./data\201801-fordgobike-tripdata.csv columns is identical
./data\201802-fordgobike-tripdata.csv columns is identical
./data\201803-fordgobike-tripdata.csv columns is identical
./data\201804-fordgobike-tripdata.csv columns is identical
./data\201805-fordgobike-tripdata.csv columns is identical
./data\201806-fordgobike-tripdata.csv columns is identical
./data\201807-fordgobike-tripdata.csv columns is identical
./data\201808-fordgobike-tripdata.csv columns is identical
./data\201809-fordgobike-tripdata.csv columns is identical
./data\201810-fordgobike-tripdata.csv columns is identical
./data\201811-fordgobike-tripdata.csv columns is identical
./data\201812-fordgobike-tripdata.csv columns is identical


In [68]:
# Concat all 2018 dataframe
df_2018 = pd.concat(map(pd.read_csv, glob.glob('data/2018*.csv')), ignore_index=True)
df_2018

duration_sec                start_time                  end_time  \
0               75284  2018-01-31 22:52:35.2390  2018-02-01 19:47:19.8240   
1               85422  2018-01-31 16:13:34.3510  2018-02-01 15:57:17.3100   
2               71576  2018-01-31 14:23:55.8890  2018-02-01 10:16:52.1160   
3               61076  2018-01-31 14:53:23.5620  2018-02-01 07:51:20.5000   
4               39966  2018-01-31 19:52:24.6670  2018-02-01 06:58:31.0530   
...               ...                       ...                       ...   
1863716           473  2018-12-01 00:11:54.8110  2018-12-01 00:19:48.5470   
1863717           841  2018-12-01 00:02:48.7260  2018-12-01 00:16:49.7660   
1863718           260  2018-12-01 00:05:27.6150  2018-12-01 00:09:47.9560   
1863719           292  2018-12-01 00:03:06.5490  2018-12-01 00:07:59.0800   
1863720           150  2018-12-01 00:03:05.7420  2018-12-01 00:05:36.0260   

         start_station_id                                 start_station_name  \
0                   120.0                               Mission Dolores Park   
1                    15.0  San Francisco Ferry Building (Harry Bridges Pl...   
2                   304.0                               Jackson St at 5th St   
3                    75.0                           Market St at Franklin St   
4                    74.0                              Laguna St at Hayes St   
...                   ...                                                ...   
1863716             345.0                              Hubbell St at 16th St   
1863717              10.0                         Washington St at Kearny St   
1863718             245.0                             Downtown Berkeley BART   
1863719              93.0                       4th St at Mission Bay Blvd S   
1863720             107.0                              17th St at Dolores St   

         start_station_latitude  start_station_longitude  end_station_id  \
0                     37.761420              -122.426435           285.0   
1                     37.795392              -122.394203            15.0   
2                     37.348759              -121.894798           296.0   
3                     37.773793              -122.421239            47.0   
4                     37.776435              -122.426244            19.0   
...                         ...                      ...             ...   
1863716               37.766474              -122.398295            81.0   
1863717               37.795393              -122.404770            58.0   
1863718               37.870139              -122.268422           255.0   
1863719               37.770407              -122.391198           126.0   
1863720               37.763015              -122.426497           119.0   

                                          end_station_name  \
0                               Webster St at O'Farrell St   
1        San Francisco Ferry Building (Harry Bridges Pl...   
2                                    5th St at Virginia St   
3                                    4th St at Harrison St   
4                                     Post St at Kearny St   
...                                                    ...   
1863716                                 Berry St at 4th St   
1863717                               Market St at 10th St   
1863718                        Virginia St at Shattuck Ave   
1863719                                        Esprit Park   
1863720                                  18th St at Noe St   

         end_station_latitude  end_station_longitude  bike_id   user_type  \
0                   37.783521            -122.431158     2765  Subscriber   
1                   37.795392            -122.394203     2815    Customer   
2                   37.325998            -121.877120     3039    Customer   
3                   37.780955            -122.399749      321    Customer   
4                   37.788975            -122.403452      617  Subscriber   
...

In [69]:
df_2018.info(verbose=True, null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1863721 entries, 0 to 1863720
Data columns (total 14 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   duration_sec             1863721 non-null  int64  
 1   start_time               1863721 non-null  object 
 2   end_time                 1863721 non-null  object 
 3   start_station_id         1851950 non-null  float64
 4   start_station_name       1851950 non-null  object 
 5   start_station_latitude   1863721 non-null  float64
 6   start_station_longitude  1863721 non-null  float64
 7   end_station_id           1851950 non-null  float64
 8   end_station_name         1851950 non-null  object 
 9   end_station_latitude     1863721 non-null  float64
 10  end_station_longitude    1863721 non-null  float64
 11  bike_id                  1863721 non-null  int64  
 12  user_type                1863721 non-null  object 
 13  bike_share_for_all_trip  1863721 non-null 

In [70]:
df_2018.isna().sum()

duration_sec                   0
start_time                     0
end_time                       0
start_station_id           11771
start_station_name         11771
start_station_latitude         0
start_station_longitude        0
end_station_id             11771
end_station_name           11771
end_station_latitude           0
end_station_longitude          0
bike_id                        0
user_type                      0
bike_share_for_all_trip        0
dtype: int64

In [71]:
df_2018[df_2018.isnull().any(axis='columns')]

duration_sec                start_time                  end_time  \
623364           5572  2018-06-30 20:58:31.8550  2018-06-30 22:31:24.5930   
623376            573  2018-06-30 22:12:27.5110  2018-06-30 22:22:01.1740   
623393            850  2018-06-30 22:01:25.5850  2018-06-30 22:15:36.1510   
623429            562  2018-06-30 21:52:05.1180  2018-06-30 22:01:27.4400   
623456            819  2018-06-30 21:37:34.5060  2018-06-30 21:51:13.7290   
...               ...                       ...                       ...   
1860854           663  2018-12-02 09:47:44.1160  2018-12-02 09:58:47.1620   
1861409          1141  2018-12-01 21:22:41.6460  2018-12-01 21:41:42.7270   
1861570           753  2018-12-01 19:57:14.2010  2018-12-01 20:09:47.8580   
1862625          2244  2018-12-01 14:01:06.2520  2018-12-01 14:38:30.8950   
1863272          1269  2018-12-01 10:54:13.7640  2018-12-01 11:15:23.0360   

         start_station_id start_station_name  start_station_latitude  \
623364                NaN                NaN                   37.40   
623376                NaN                NaN                   37.40   
623393                NaN                NaN                   37.41   
623429                NaN                NaN                   37.41   
623456                NaN                NaN                   37.41   
...                   ...                ...                     ...   
1860854               NaN                NaN                   37.40   
1861409               NaN                NaN                   37.41   
1861570               NaN                NaN                   37.41   
1862625               NaN                NaN                   37.41   
1863272               NaN                NaN                   37.42   

         start_station_longitude  end_station_id end_station_name  \
623364                   -121.94             NaN              NaN   
623376                   -121.94             NaN              NaN   
623393                   -121.94             NaN              NaN   
623429                   -121.94             NaN              NaN   
623456                   -121.94             NaN              NaN   
...                          ...             ...              ...   
1860854                  -121.92             NaN              NaN   
1861409                  -121.96             NaN              NaN   
1861570                  -121.94             NaN              NaN   
1862625                  -121.93             NaN              NaN   
1863272                  -121.96             NaN              NaN   

         end_station_latitude  end_station_longitude  bike_id   user_type  \
623364                  37.40                -121.94     4202    Customer   
623376                  37.40                -121.93     4095  Subscriber   
623393                  37.41                -121.95     4122    Customer   
623429                  37.42                -121.94     4184  Subscriber   
623456                  37.41                -121.96     4137    Customer   
...                       ...                    ...      ...         ...   
1860854                 37.39                -121.93     4284    Customer   
1861409                 37.41                -121.94     4256  Subscriber   
1861570                 37.40                -121.93     4181  Subscriber   
1862625                 37.42                -121.96     4238    Customer   
1863272                 37.40                -121.92     4284    Customer   

        bike_share_for_all_trip  
623364                       No  
623376                      Yes  
623393                       No  
623429                       No  
623456                       No  
...                         ...  
1860854                      No  
1861409                      No  
1861570                      No  
1862625                      No  
1863272                      No  

[11771 rows x 14 columns]

### 2019 Dataframes

In [77]:
# Column check for year 2019, starterd from index 13 to 24
column_check(13)

./data\201901-fordgobike-tripdata.csv columns is identical
./data\201902-fordgobike-tripdata.csv columns is identical
./data\201903-fordgobike-tripdata.csv columns is identical
./data\201904-fordgobike-tripdata.csv columns is identical
./data\201905-baywheels-tripdata.csv columns is identical
./data\201906-baywheels-tripdata.csv columns is NOT identical
./data\201907-baywheels-tripdata.csv columns is NOT identical
./data\201908-baywheels-tripdata.csv columns is identical
./data\201909-baywheels-tripdata.csv columns is identical
./data\201910-baywheels-tripdata.csv columns is identical
./data\201911-baywheels-tripdata.csv columns is NOT identical
./data\201912-baywheels-tripdata.csv columns is NOT identical


In [81]:
# Check difference in column
print(pd.read_csv(csv_list[13]).columns)
print(pd.read_csv(csv_list[18]).columns)
print(pd.read_csv(csv_list[19]).columns)
print(pd.read_csv(csv_list[23]).columns)
print(pd.read_csv(csv_list[24]).columns)

Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type',
       'bike_share_for_all_trip'],
      dtype='object')


C:\Users\muham\miniconda3\envs\dan_udacity\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (14) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type',
       'bike_share_for_all_trip', 'rental_access_method'],
      dtype='object')


C:\Users\muham\miniconda3\envs\dan_udacity\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (13,14) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type',
       'bike_share_for_all_trip', 'rental_access_method'],
      dtype='object')
Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type',
       'bike_share_for_all_trip', 'rental_access_method'],
      dtype='object')
Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type',
       're

C:\Users\muham\miniconda3\envs\dan_udacity\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (13) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [79]:
df_201901 = pd.read_csv(csv_list[13])
df_201901               

duration_sec                start_time                  end_time  \
0              80825  2019-01-31 17:57:44.6130  2019-02-01 16:24:49.8640   
1              65900  2019-01-31 20:58:33.8860  2019-02-01 15:16:54.1730   
2              62633  2019-01-31 18:06:52.9240  2019-02-01 11:30:46.5300   
3              44680  2019-01-31 19:46:09.7190  2019-02-01 08:10:50.3180   
4              60709  2019-01-31 14:19:01.5410  2019-02-01 07:10:51.0650   
...              ...                       ...                       ...   
192077           390  2019-01-01 00:16:36.8450  2019-01-01 00:23:07.2530   
192078           227  2019-01-01 00:17:40.9750  2019-01-01 00:21:28.4180   
192079           565  2019-01-01 00:11:03.4410  2019-01-01 00:20:29.2680   
192080           182  2019-01-01 00:09:56.2110  2019-01-01 00:12:58.7480   
192081           226  2019-01-01 00:07:10.5760  2019-01-01 00:10:57.1960   

        start_station_id            start_station_name  \
0                  229.0     Foothill Blvd at 42nd Ave   
1                    4.0   Cyril Magnin St at Ellis St   
2                  245.0        Downtown Berkeley BART   
3                   85.0       Church St at Duboce Ave   
4                   16.0       Steuart St at Market St   
...                  ...                           ...   
192077             109.0        17th St at Valencia St   
192078              73.0        Pierce St at Haight St   
192079             269.0  Telegraph Ave at Carleton St   
192080             269.0  Telegraph Ave at Carleton St   
192081             245.0        Downtown Berkeley BART   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.775745              -122.213037           196.0   
1                    37.785881              -122.408915           134.0   
2                    37.870139              -122.268422           157.0   
3                    37.770083              -122.429156            53.0   
4                    37.794130              -122.394430            28.0   
...                        ...                      ...             ...   
192077               37.763316              -122.421904           141.0   
192078               37.771793              -122.433708            96.0   
192079               37.862320              -122.258801           243.0   
192080               37.862320              -122.258801           240.0   
192081               37.870139              -122.268422           247.0   

                      end_station_name  end_station_latitude  \
0              Grand Ave at Perkins St             37.808894   
1               Valencia St at 24th St             37.752428   
2                 65th St at Hollis St             37.846784   
3               Grove St at Divisadero             37.775946   
4         The Embarcadero at Bryant St             37.787168   
...                                ...                   ...   
192077  Valencia St at Cesar Chavez St             37.747998   
192078           Dolores St at 15th St             37.766210   
192079     Bancroft Way at College Ave             37.869360   
192080       Haste St at Telegraph Ave             37.866043   
192081       Fulton St at Bancroft Way             37.867789   

        end_station_longitude  bike_id   user_type bike_share_for_all_trip  
0                 -122.256460     4861  Subscriber                      No  
1                 -122.420628     5506  Subscriber                      No  
2                 -122.291376     2717    Customer                      No  
3                 -122.437777     4557    Customer                      No  
4                 -122.388098     2100    Customer                      No  
...                       ...      ...         ...                     ...  
192077            -122.420219     5059  Subscriber                      No  
192078            -122.426614     4659  Subscriber                      No  
192079            -122.254337     4

In [80]:
df_201906 = pd.read_csv(csv_list[18])
df_201906               

C:\Users\muham\miniconda3\envs\dan_udacity\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (14) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


duration_sec                start_time                  end_time  \
0              81695  2019-06-30 18:16:09.7730  2019-07-01 16:57:45.5920   
1              74260  2019-06-30 18:09:55.8300  2019-07-01 14:47:36.6810   
2              59603  2019-06-30 15:40:31.0380  2019-07-01 08:13:54.3490   
3              67932  2019-06-30 12:49:47.7810  2019-07-01 07:42:00.5430   
4              48894  2019-06-30 17:21:00.0550  2019-07-01 06:55:54.9960   
...              ...                       ...                       ...   
191767          1453       2019-06-23 18:40:11       2019-06-23 19:04:24   
191768          1674       2019-06-24 22:07:37       2019-06-24 22:35:32   
191769           672       2019-06-26 16:07:51       2019-06-26 16:19:04   
191770          1385       2019-06-26 22:34:47       2019-06-26 22:57:53   
191771           794       2019-06-24 20:34:47       2019-06-24 20:48:02   

        start_station_id                                 start_station_name  \
0                  109.0                             17th St at Valencia St   
1                   50.0                              2nd St at Townsend St   
2                   23.0                      The Embarcadero at Steuart St   
3                  169.0                                       Bushrod Park   
4                   15.0  San Francisco Ferry Building (Harry Bridges Pl...   
...                  ...                                                ...   
191767               NaN                                                NaN   
191768               NaN                                                NaN   
191769               NaN                                                NaN   
191770               NaN                                                NaN   
191771               NaN                                                NaN   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.763316              -122.421904            56.0   
1                    37.780526              -122.390288           101.0   
2                    37.791464              -122.391034            30.0   
3                    37.846516              -122.265304           248.0   
4                    37.795392              -122.394203            30.0   
...                        ...                      ...             ...   
191767               37.416979              -121.956381             NaN   
191768               37.417026              -121.954276             NaN   
191769               37.417792              -121.955901             NaN   
191770               37.417983              -121.955682             NaN   
191771               37.418230              -121.955282             NaN   

                                      end_station_name  end_station_latitude  \
0                                        Koshland Park             37.773414   
1                               15th St at Potrero Ave             37.767079   
2       San Francisco Caltrain (Townsend St at 4th St)             37.776598   
3                           Telegraph Ave at Ashby Ave             37.855956   
4       San Francisco Caltrain (Townsend St at 4th St)             37.776598   
...                                                ...                   ...   
191767                                             NaN             37.396961   
191768                                             NaN             37.406156   
191769                                             NaN             37.413978   
191770                                             NaN             37.405492   
191771                                             NaN             37.405155   

        end_station_longitude  bike_id   user_type bike_share_for_all_trip  \
0                 -122.427317     1502  Subscriber                      No   
1                 -122.407359     2526    Customer                      No   
2                 -122.395282     2427  Subscriber

###  2020 Dataframes

In [45]:
# Column check for year 2020, starterd from index 25 to the end
column_check(25)

./data\202001-baywheels-tripdata.csv columns is same
./data\202002-baywheels-tripdata.csv columns is same
./data\202003-baywheels-tripdata.csv columns is same
different
different
different
different


## Data Assesment Result

#### Year 2018
* 11771 null value based on start_station_id, start_station_name, end_station_id, and end_station_name columns
* bike_share_for_all_trip

#### Year 2019
* Columns name is not same for all csv files

#### Year 2020
* Columns name is not same for all csv files


# Data Cleaning